File: water_meter.rb
# water_meter.rb
#!/usr/bin/env ruby
require 'json'
require 'date'
require 'optparse'

DATA_FILE = 'water_data.json'

class WaterMeter
  attr_reader :data

  def initialize
    @data = []
    load
  end

  def load
    if File.exist?(DATA_FILE)
      begin
        @data = JSON.parse(File.read(DATA_FILE))
      rescue
        @data = []
      end
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@data))
  end

  def add(value, unit = 'L', date = nil)
    date ||= Date.today.to_s
    @data << { 'date' => date, 'value' => value, 'unit' => unit }
    save
    puts "✅ Added: #{value} #{unit} on #{date}"
  end

  def list
    if @data.empty?
      puts "No data."
      return
    end
    puts "\n📋 All readings:"
    @data.each do |e|
      puts "  #{e['date']}: #{e['value']} #{e['unit']}"
    end
  end

  def stats(days = 7, unit = nil)
    if @data.empty?
      puts "No data."
      return
    end
    cutoff = Date.today - days
    filtered = @data.select { |e| Date.parse(e['date']) >= cutoff }
    if filtered.empty?
      puts "No data in the last #{days} days."
      return
    end
    target_unit = unit || 'L'
    total = 0
    filtered.each do |e|
      val = e['value']
      if e['unit'] == 'm3' && target_unit == 'L'
        val *= 1000
      elsif e['unit'] == 'L' && target_unit == 'm3'
        val /= 1000
      end
      total += val
    end
    avg = total / filtered.length
    puts "\n📊 Statistics (last #{days} days) – unit: #{target_unit}"
    puts "  Total: %.2f #{target_unit}" % total
    puts "  Average per day: %.2f #{target_unit}" % avg
  end

  def reset
    @data = []
    save
    puts "All data cleared."
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: water_meter.rb <command> [options]"
  opts.on("add VALUE", Float, "Add a reading (default unit L)") { |v| options[:add_value] = v }
  opts.on("--unit UNIT", "L or m3") { |v| options[:unit] = v }
  opts.on("--date DATE", "YYYY-MM-DD") { |v| options[:date] = v }
  opts.on("list", "List all readings") { options[:cmd] = :list }
  opts.on("stats", "Show statistics") { options[:cmd] = :stats }
  opts.on("--days N", Integer, "Days for stats") { |v| options[:days] = v }
  opts.on("--stats-unit UNIT", "L or m3") { |v| options[:stats_unit] = v }
  opts.on("reset", "Clear all data") { options[:cmd] = :reset }
end.parse!

app = WaterMeter.new

if options[:add_value]
  app.add(options[:add_value], options[:unit] || 'L', options[:date])
elsif options[:cmd] == :list
  app.list
elsif options[:cmd] == :stats
  app.stats(options[:days] || 7, options[:stats_unit])
elsif options[:cmd] == :reset
  app.reset
else
  puts "Unknown command. Use add, list, stats, reset."
end

File: water_meter.php
# water_meter.php
#!/usr/bin/env php
<?php

$dataFile = 'water_data.json';

class WaterMeter {
    private $data = [];
    private $file;

    function __construct($file) {
        $this->file = $file;
        $this->load();
    }

    function load() {
        if (file_exists($this->file)) {
            $content = file_get_contents($this->file);
            $this->data = json_decode($content, true) ?? [];
        }
    }

    function save() {
        file_put_contents($this->file, json_encode($this->data, JSON_PRETTY_PRINT));
    }

    function add($value, $unit = 'L', $date = null) {
        if ($date === null) $date = date('Y-m-d');
        $this->data[] = ['date' => $date, 'value' => $value, 'unit' => $unit];
        $this->save();
        echo "✅ Added: $value $unit on $date\n";
    }

    function list() {
        if (empty($this->data)) {
            echo "No data.\n";
            return;
        }
        echo "\n📋 All readings:\n";
        foreach ($this->data as $e) {
            echo "  {$e['date']}: {$e['value']} {$e['unit']}\n";
        }
    }

    function stats($days = 7, $unit = null) {
        if (empty($this->data)) {
            echo "No data.\n";
            return;
        }
        $cutoff = new DateTime();
        $cutoff->modify("-$days days");
        $filtered = array_filter($this->data, function($e) use ($cutoff) {
            return new DateTime($e['date']) >= $cutoff;
        });
        if (empty($filtered)) {
            echo "No data in the last $days days.\n";
            return;
        }
        $targetUnit = $unit ?? 'L';
        $total = 0;
        foreach ($filtered as $e) {
            $val = $e['value'];
            if ($e['unit'] == 'm3' && $targetUnit == 'L') $val *= 1000;
            elseif ($e['unit'] == 'L' && $targetUnit == 'm3') $val /= 1000;
            $total += $val;
        }
        $avg = $total / count($filtered);
        echo "\n📊 Statistics (last $days days) – unit: $targetUnit\n";
        echo "  Total: " . number_format($total, 2) . " $targetUnit\n";
        echo "  Average per day: " . number_format($avg, 2) . " $targetUnit\n";
    }

    function reset() {
        $this->data = [];
        $this->save();
        echo "All data cleared.\n";
    }
}

$app = new WaterMeter($dataFile);

$cmd = $argv[1] ?? null;
if ($cmd === 'add') {
    $value = (float)($argv[2] ?? 0);
    $unit = $argv[3] ?? 'L';
    $date = null;
    for ($i=4; $i<$argc; $i++) {
        if ($argv[$i] == '--date' && isset($argv[$i+1])) {
            $date = $argv[$i+1];
            $i++;
        }
    }
    $app->add($value, $unit, $date);
} elseif ($cmd === 'list') {
    $app->list();
} elseif ($cmd === 'stats') {
    $days = 7;
    $unit = null;
    for ($i=2; $i<$argc; $i++) {
        if ($argv[$i] == '--days' && isset($argv[$i+1])) {
            $days = (int)$argv[$i+1];
            $i++;
        }
        if ($argv[$i] == '--unit' && isset($argv[$i+1])) {
            $unit = $argv[$i+1];
            $i++;
        }
    }
    $app->stats($days, $unit);
} elseif ($cmd === 'reset') {
    $app->reset();
} else {
    echo "Usage: water_meter.php add <value> [unit] [--date YYYY-MM-DD] | list | stats [--days N] [--unit L|m3] | reset\n";
}
?>
